pkg/controller/certificatemanager/certificatemanager.go

```diff
@@ -0,0 +1,286 @@
+package certificatemanager
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/openshift/library-go/pkg/crypto"
+	operatorv1 "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/common"
+	"github.com/tigera/operator/pkg/components"
+	"github.com/tigera/operator/pkg/controller/status"
+	"github.com/tigera/operator/pkg/controller/utils/imageset"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/tls"
+	"github.com/tigera/operator/pkg/tls/certificatemanagement"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var (
+	log                    = logf.Log.WithName("tls")
+	ErrInvalidCertDNSNames = errors.New("certificate has the wrong DNS names")
+	errNoPrivateKeyPEM     = errors.New("key pair is missing a private key")
+	errNoCertificatePEM    = errors.New("certificate PEM is missing")
+	errCreatedByOldCA      = errors.New("certificate was created by an older ca")
+)
+
+type certificateManager struct {
+	*x509.Certificate
+	*crypto.CA
+	keyPair *certificatemanagement.KeyPair
+}
+
+// CertificateManager can sign new certificates and has methods to retrieve existing KeyPairs and Certificates. If a user
+// brings their own secrets, CertificateManager will preserve and return them.
+type CertificateManager interface {
+	// GetKeyPair returns an existing KeyPair. If the KeyPair is not found, nil is returned.
+	GetKeyPair(cli client.Client, secretName, secretNamespace string) (certificatemanagement.KeyPairInterface, error)
+	// GetOrCreateKeyPair returns a KeyPair. If one exists, some checks are performed. Otherwise, a new KeyPair is created.
+	GetOrCreateKeyPair(cli client.Client, secretName, secretNamespace string, dnsNames []string) (certificatemanagement.KeyPairInterface, error)
+	// GetCertificate returns a Certificate. If the certificate is not found, nil is returned.
+	GetCertificate(cli client.Client, secretName, secretNamespace string) (certificatemanagement.CertificateInterface, error)
+	// AddToStatusManager lets the status manager monitor pending CSRs if the certificate management is enabled.
+	AddToStatusManager(manager status.StatusManager, namespace string)
+	KeyPair() certificatemanagement.KeyPairInterface
```

Comment:
I can't remember if you had this before and changed to doing this (possibly because of comment from me), but seeing that this is only ever used to populate `certificatemanagement.CreateTrustedBundle` it might be more intuitive if `CertificateManager` instead have the `CreateTrustedBundle` function instead of this.

This function seems out of place, since this is the Manager I would think we wouldn't need to expose certs / keys inside it, and just let the manager do what it needs to do (i.e. create the trusted bundle).

If you don't replace this with the `CreateTrustedBundle` function then you should at the very least explain what this is (not just a KeyPair but the root key pair for everything this cert manager created) and explain how this should be used to create a trusted bundle. I think this comment might be a little convaluted and I'd prefer replacing this function, but I won't hold this against approving this PR.